stan/math/prim/arr/meta/scalar_type.hpp

```diff
@@ -2,27 +2,20 @@
 #define STAN_MATH_PRIM_ARR_META_SCALAR_TYPE_HPP
 
 #include <stan/math/prim/scal/meta/scalar_type.hpp>
+#include <stan/math/prim/arr/meta/is_vector.hpp>
+#include <type_traits>
 #include <vector>
 
 namespace stan {
-template <typename T>
-struct scalar_type<std::vector<T> > {
-  typedef typename scalar_type<T>::type type;
-};
 
+/**
+ * Return the underlying type from a standard vector.
+ * @param T The standard vector to extract the underlying type from.
+ */
 template <typename T>
-struct scalar_type<const std::vector<T> > {
-  typedef typename scalar_type<T>::type type;
+struct scalar_type<T, std::enable_if_t<is_std_vector<T>::value>> {
+  typedef typename scalar_type<typename T::value_type>::type type;
```

Comment:
fyi this uses `std::decay<>` so we don't have to have the const and ref versions of this functions. I like this more because we do the same things for const/refs etc.